x-pack/platform/plugins/shared/fleet/server/tasks/upgrade_agentless_deployment.test.ts

```diff
@@ -0,0 +1,344 @@
+          },
+        ],
+      });
+      await runTask();
+
+      expect(mockAgentPolicyService.fetchAllAgentPolicies).toHaveBeenCalled();
+      expect(agentlessAgentService.upgradeAgentlessDeployment).not.toHaveBeenCalled();
+    });
+
+    it('should upgrade agentless deployments when agent status is online', async () => {
+      mockedGetLatestAvailableAgentVersion.mockResolvedValue('8.17.1');
+      mockedGetAgentsByKuery.mockResolvedValue({
+        agents: [
+          {
+            id: 'agent-1',
+            policy_id: '93c46720-c217-11ea-9906-b5b8a21b268e',
+            status: 'online',
+            agent: {
+              version: '8.17.0',
+            },
+          },
+        ],
+      });
+      await runTask();
+
+      expect(mockAgentPolicyService.fetchAllAgentPolicies).toHaveBeenCalled();
+      expect(agentlessAgentService.upgradeAgentlessDeployment).toHaveBeenCalled();
+    });
+
+    it('should not upgrade agentless deployments when agent status is unenroll', async () => {
+      mockedGetLatestAvailableAgentVersion.mockResolvedValue('8.17.1');
+      mockedGetAgentsByKuery.mockResolvedValue({
+        agents: [
+          {
+            id: 'agent-1',
+            policy_id: '93c46720-c217-11ea-9906-b5b8a21b268e',
+            status: 'unenroll',
+            agent: {
+              version: '8.17.0',
+            },
+          },
+        ],
+      });
+      await runTask();
+
+      expect(mockAgentPolicyService.fetchAllAgentPolicies).toHaveBeenCalled();
+      expect(agentlessAgentService.upgradeAgentlessDeployment).not.toHaveBeenCalled();
+    });
+
+    it('should upgrade agentless deployments when agent version is up to date', async () => {
+      mockedGetLatestAvailableAgentVersion.mockResolvedValue('8.17.1');
+      mockedGetAgentsByKuery.mockResolvedValue({
+        agents: [
+          {
+            id: 'agent-1',
+            policy_id: '93c46720-c217-11ea-9906-b5b8a21b268e',
+            status: 'online',
+            agent: {
+              version: '8.17.0',
+            },
+          },
+        ],
+      });
+      await runTask();
+
+      expect(mockAgentPolicyService.fetchAllAgentPolicies).toHaveBeenCalled();
+      expect(agentlessAgentService.upgradeAgentlessDeployment).toHaveBeenCalled();
+    });
+
+    it('should not call upgrade agentless api to upgrade when 0 agents', async () => {
+      mockedGetAgentsByKuery.mockResolvedValue({
+        agents: [],
+      });
+      await runTask();
+
+      expect(mockAgentPolicyService.fetchAllAgentPolicies).toHaveBeenCalled();
+      expect(agentlessAgentService.upgradeAgentlessDeployment).not.toHaveBeenCalled();
+    });
+
+    it('should throw an error if task is aborted', async () => {
+      mockTask.abortController = new AbortController();
+
+      mockTask.abortController.signal.throwIfAborted = jest.fn(() => {
+        throw new Error('Task aborted!');
+      });
+
+      mockTask.abortController.abort();
+      await runTask();
```

Comment:
I think this is sufficient.  It would probably be better to test the effects of the `throwIfAborted()` - to make sure the task actually stopped - but a brief look at this test setup makes me think this might not be easy.

Reply:
I did try to create  cancelTask setup but it was a bit tricky simulate the throw!